### PR TITLE
[JUJU-2310] Secret multi backends

### PR DESCRIPTION
**Pull request by wallyworld**
Add support for multiple secret backends per model.

Previously, a model used just one secret backend. Now, the backend can be changed and new revisions are stored in the new backend. Older revisions are still fetched from the backend they were stored in.

A key change is to model external secret value references to include the backend id as well as revision id. This needed to be weaved through the code.

The provider interfaces are improved to take config structs directly instead of a model interface.
Operations like model cleanup now need to iterate through all configured backends to deal with revisions in any of them

Broken up into 4 commits - state, provider, uniter, facades.

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

add a model and deploy dummy-source and dummy-sink and relate
create a secret in dummy-source
check mongo that secret content is stored there
dummy-sink cannot read secret yet
grant access to dummy-sink, check secret-get

add vault as a backend
repeat the above steps, checking secret content is now in vault

show-secret --reveal on both secrets will print content
destroy model will cleanup secret in vault

repeat above on a k8s model, except out of the box, secrets are stored in k8s not mongo

### CODE REVIEW

File: secrets/backend.go [hunk 1]
```diff
@@ -53,29 +67,37 @@ func (c *secretsClient) GetContent(uri *secrets.URI, label string, refresh, peek
...
+		return content.SecretValue, nil
 	}
-	if content.BackendId != nil {
-		return c.backend.GetContent(context.Background(), *content.BackendId)
+	backend, ok := c.backends[content.ValueRef.BackendID]
+	if !ok {
+		return nil, errors.NotFoundf("external secret backend %q", content.ValueRef.BackendID)
 	}
-	return content.SecretValue, nil
+	return backend.GetContent(context.Background(), content.ValueRef.RevisionID)
```
Comment (hpidcock) [inline secrets/backend.go:77]: should this not be context.TODO()?

File: secrets/provider/provider.go [hunk 2]
```diff
@@ -78,22 +66,23 @@ type SecretBackendProvider interface {
...
 	// CleanupModel removes any secrets / ACLs / resources
-	// associated with the model.
-	CleanupModel(m Model) error
+	// associated with the model config.
+	CleanupModel(cfg *ModelBackendConfig) error
 
-	// BackendConfig returns the config needed to create a vault secrets backend client
-	// used to manage owned secrets and read shared secrets.
-	BackendConfig(m Model, tag names.Tag, owned SecretRevisions, read SecretRevisions) (*BackendConfig, error)
+	// RestrictedConfig returns the config needed to create a vault
```
Comment (hpidcock) [inline secrets/provider/provider.go:80]: what a sentence!

File: state/migration_export.go [hunk 3]
```diff
@@ -1739,7 +1739,8 @@ func (e *exporter) secrets() error {
 			Created:    rev.CreateTime,
 			Updated:    rev.UpdateTime,
 			ExpireTime: rev.ExpireTime,
-			BackendId:  rev.BackendId,
```
Comment (hpidcock) [inline state/migration_export.go:1742]: Probably error then if this isn't implemented yet.

**Review by hpidcock — APPROVED** [review]

File: state/migration_export.go (hunk 3 above)
Comment (wallyworld) [reply state/migration_export.go:1742]: I just added it

File: secrets/backend.go (hunk 1 above)
Comment (wallyworld) [reply secrets/backend.go]: yup

**Pull request merged**